plugins/kubernetes/src/utils/clusterLinks/formatters/openshift.ts

```diff
@@ -15,10 +15,39 @@
  */
 import { ClusterLinksFormatterOptions } from '../../../types/types';
 
+const kindMappings: Record<string, string> = {
+  deployment: 'deployments',
+  ingress: 'ingresses',
+  service: 'services',
+  horizontalpodautoscaler: 'horizontalpodautoscalers',
+  persistentvolume: 'persistentvolumes',
+};
+
 export function openshiftFormatter(
-  _options: ClusterLinksFormatterOptions,
-): URL {
-  throw new Error(
-    'OpenShift formatter is not yet implemented. Please, contribute!',
-  );
+  options: ClusterLinksFormatterOptions,
+): string {
+  const result = new URL(options.dashboardUrl.href);
+  const name = options.object.metadata?.name;
+  const namespace = options.object.metadata?.namespace;
+  const validKind = kindMappings[options.kind.toLocaleLowerCase('en-US')];
+  const segments = ['k8s'];
+  if (namespace) {
+    if (name && validKind) {
+      segments.push('ns', namespace, validKind, name);
+    } else {
+      segments.push('cluster', 'projects', namespace);
+    }
+  } else if (validKind) {
+    segments.push('cluster', validKind);
+    if (name) {
+      segments.push(name);
+    }
+  }
+  if (segments.length > 1) {
+    // define a path only when we have more segments than the initial 'k8s'
+    result.pathname = segments
```

Comment:
So what happens here is you replace the entire path unconditionally. What if your base URL is `https://example.com/x/openshift-web`? Seems to me that you should have kept the URL return type, and used the relative resolution constructor of URL.

```ts
const name = encodeURIComponent(options.object.metadata?.name ?? '');
const namespace = encodeURIComponent(options.object.metadata?.namespace ?? '');
const validKind = kindMappings[options.kind.toLocaleLowerCase('en-US')];

// this works well with the constructor on the last line, so you can also
// skip the segments.length === 1 check
let path = ''; 
if (namespace) {
  if (name && validKind) {
    path = `k8s/ns/${namespace}/${validKind}/${name}`;
  }
  // ...etc

return new URL(path, options.dashboardUrl.href);
```

Side note: maybe the validKinds thing could be replaced with a `pluralizeKind` function that just returns the input + `'s'` except in special circumstances? :P That would make it work magically for most english words without extra effort. And then that could be made into a helper function that's shared for the entire project later on

Reply:
I rolled back my changes and refactored the code with your suggestions.
I applied them to the other formatters and I moved the special case of URL rendering into the standardFormatter which needs url fragments, but not the other ones.
I think this is cleaner that way.